src/mpid/ch4/src/ch4r_rma_target_callbacks.c

```diff
@@ -1655,27 +1646,27 @@ int MPIDIG_put_data_target_msg_cb(int handler_id, void *am_hdr, void *data, MPI_
     int mpi_errno = MPI_SUCCESS;
     MPIR_Request *rreq;
     MPIDIG_put_dat_msg_t *msg_hdr = (MPIDIG_put_dat_msg_t *) am_hdr;
-    struct iovec *iov;
-    uintptr_t base;
-    int i;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_PUT_DATA_TARGET_MSG_CB);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_PUT_DATA_TARGET_MSG_CB);
     MPIR_T_PVAR_TIMER_START(RMA, rma_targetcb_put_data);
 
     rreq = (MPIR_Request *) msg_hdr->preq_ptr;
-    base = (uintptr_t) MPIDIG_REQUEST(rreq, req->preq.target_addr);
 
-    /* Adjust the target iov addresses using the base address
-     * (window base + target_disp) */
-    iov = (struct iovec *) MPIDIG_REQUEST(rreq, req->preq.dt_iov);
-    for (i = 0; i < MPIDIG_REQUEST(rreq, req->preq.n_iov); i++)
-        iov[i].iov_base = (char *) iov[i].iov_base + base;
+    /* FIXME: MPIR_Typerep_unflatten should allocate the new object */
+    MPIR_Datatype *dt = (MPIR_Datatype *) MPIR_Handle_obj_alloc(&MPIR_Datatype_mem);
+    if (!dt) {
+        MPIR_ERR_SETANDJUMP1(mpi_errno, MPI_ERR_OTHER, "**nomem", "**nomem %s",
+                             "MPIR_Datatype_mem");
+    }
+    /* Note: handle is filled in by MPIR_Handle_obj_alloc() */
+    MPIR_Object_set_ref(dt, 1);
+    MPIR_Typerep_unflatten(dt, MPIDIG_REQUEST(rreq, req->preq.flattened_dt));
+    MPIDIG_REQUEST(rreq, datatype) = dt->handle;
+    MPIDIG_REQUEST(rreq, count) = in_data_sz / dt->size;
 
-    void *dt_iov = MPIDIG_REQUEST(rreq, req->preq.dt_iov);
-    int n_iov = MPIDIG_REQUEST(rreq, req->preq.n_iov);
     MPIDIG_REQUEST(rreq, req->target_cmpl_cb) = put_target_cmpl_cb;
-    MPIDIG_recv_init(0, in_data_sz, dt_iov, n_iov, rreq);
+    MPIDIG_recv_type_init(in_data_sz, rreq);
```

Comment:
`MPIDIG_recv_type_init` currently refers to `MPI_Datatype`, but it is trivial to add an enum (refer to `mpidig_msg.h`) to directly use `MPIR_Datatype *` in stead.